File: ComputationalStatisticsInPython/FinancialRisk_VaR.py
################################
#
# Understanding and Applying Financial Risk Modeling Techniques
# https://app.pluralsight.com/library/courses/financial-risk-modeling-techniques/table-of-contents
# Implementing Financial Risk Models in R
#
#################################


#################################
# Load libraries
#################################

# changing directory
import os

# data frames
import pandas as pd

# statistcial work
import numpy as np

# for factor models
import statsmodels.api as sm


#################################
# Load and clean data
#################################

# Set working directory
os.chdir("../Data")
workingDirectory = os.getcwd()
workingDirectory

# Get data
portfoliosFilePath = workingDirectory + "/Portfolios.csv"
portfolios = pd.read_csv(portfoliosFilePath, sep=",")
portfolios.head()

# format date
portfolios['AsOfDate'] = pd.to_datetime(portfolios['AsOfDate'], format='%m/%d/%Y', yearfirst = True)
portfolios.head()

# sort by date
portfolios = portfolios.sort_values(['AsOfDate'] , ascending=True)
portfolios.head()


#################################
# Estimating Historical Risk
#################################

# calculate returns
returns = portfolios[[ key for key in dict(portfolios.dtypes) if dict(portfolios.dtypes)[key] in [ 'float64', 'int64']]].pct_change()
returns = returns[1:]
returns.head()

# add intercept column
returns['Intercept'] = 1
returns.head()

# create lists of names to easily filter returns
stockNames = list(returns)[0:10]
factorNames = list(returns)[10:13]

# set values for returns and weights
stockReturns = returns[stockNames]
factorReturns = returns[factorNames]
weights = np.array([1.0/len(stockNames)]*len(stockNames))


# cacluate portfolio variance
# Var(p) = W * COV(Y) * W_transposed
# matrix multiplication is done via np.dot
# first is transpose of weights
# second is the covariance of the returns
# third is the matrix of weights
historicalRisk = np.dot(np.dot(weights, stockReturns.cov()), weights.T)



#################################
# Building Factor Models
#################################

# calculate lm (regression) for FVX, SP500, and the stock value

# make set to use
factorData = factorReturns
factorData.head()

# loops for each stock to create
modelCoeffs =[]
for stockName in stockNames:
    stockReturn = returns[stockName]
    model = sm.OLS(stockReturn, factorData)
    result = model.fit()
    modelCoeffRow = list(result.params)
    modelCoeffRow.append(np.std(result.resid,ddof=1))
    modelCoeffs.append(modelCoeffRow)
    #print(result.summary())
    
# convert and cleanup data frame
modelCoeffs = pd.DataFrame(modelCoeffs)
modelCoeffs.columns = ["B_FVX", "B_SP500",  "Alpha", "ResidualVol"]
modelCoeffs["Names"] = stockNames
modelCoeffs


#################################
# Factor Analysis - Idiosyncratic and Systemic Risk
#################################

# TotalVaR(P) = SystemicVaR(P) + IdisyncraticVaR(P)

# SystemicVaR(P) = weights * betas * Cov(Factors) * betas_transposed * weights_transposed
# inner terms
factorCovariance = factorReturns[["SP500","FVX"]].cov()
reconstructedCov = np.dot(np.dot(modelCoeffs[["B_SP500", "B_FVX"]],factorCovariance),modelCoeffs[["B_SP500", "B_FVX"]].T)

# include outer terms
systemicVariance = np.dot(np.dot(weights, reconstructedCov), weights.T)

# idosyncraticVariance(P) = weights * var(residuals) * weight_transposed
idiosyncraticVariance = sum(weights * modelCoeffs["ResidualVol"] * weights * modelCoeffs["ResidualVol"])

# totalVariance = systemicVariance + idiosyncraticVariance
factorBasedVariance = systemicVariance + idiosyncraticVariance


#################################
# Scenario-based Stress Testing
#################################

# create a range of scenarios, stepping from min to max for factors, FVX and S&P500
fvxScenarios = np.arange(min(returns["FVX"]), max(returns["FVX"]), .05)
sp500Scenarios = np.arange(min(returns["SP500"]), max(returns["SP500"]), .02)

# build scenarios from fvxScenarios by sp500Scenarios
scenarios = []
for fvxValue in fvxScenarios:
    for sp500value in sp500Scenarios:
        scenario = [fvxValue, sp500value]
        for stockName in stockNames:
            alpha = float(modelCoeffs[modelCoeffs["Names"] == stockName]["Alpha"])
            beta_sp = float(modelCoeffs[modelCoeffs["Names"] == stockName]["B_SP500"])
            beta_fvx = float(modelCoeffs[modelCoeffs["Names"] == stockName]["B_FVX"])
            scenarioPredictedReturn = alpha + (beta_sp * sp500value) + (beta_fvx * fvxValue)
            scenario.append(scenarioPredictedReturn)
        scenarios.append(scenario)

scenarios = pd.DataFrame(scenarios)

# set column names
scenarios.columns = ["FVX","SP500","AAPL","ADBE","CVX","GOOG","IBM","MDLZ","MSFT","NFLX","ORCL","SBUX"]
scenarios.head()

# totalVariance(P) = weights * cov(scenarios) * weight_transposed
scenariosCov = scenarios[stockNames].cov()
scenarioTotalVariance = np.dot(np.dot(weights, scenariosCov), weights.T)
scenarioTotalVariance


#################################
# Quantifying the Worst Case
#################################

# VaR = P x Z_alpha x stdDev
def calculateVaR(risk, confidenceLevel, principal = 1, numMonths = 1):
    vol = math.sqrt(risk)
    return abs(principal * norm.ppf(1-confidenceLevel, 0, 1) * vol * math.sqrt(numMonths))

# Worst Case: VaR based on scenario-based stress testing
calculateVaR(scenarioTotalVariance, 0.99)

# Allternaitve Risk Measure:  VaR based on factors
calculateVaR(factorBasedVariance, 0.99)

# Allternaitve Risk Measure:  VaR based on history
calculateVaR(historicalRisk, 0.99)

File: RTutorials/RTutorials/FinancialRisk_VaR.R
################################
#
# Understanding and Applying Financial Risk Modeling Techniques
# https://app.pluralsight.com/library/courses/financial-risk-modeling-techniques/table-of-contents
# Implementing Financial Risk Models in R
#
#################################


#################################
# Load and clean data
#################################

# set WD
getwd()
#setwd("../Data")
#getwd()

# load data
filePath <- "Portfolios.csv"
portfolios.dirty <- read.table(filePath, header = TRUE, sep = ",", stringsAsFactors = FALSE)

# convert: http://www.statmethods.net/input/dates.html
portfolios.dirty[, c("AsOfDate")] <- as.Date(portfolios.dirty[, c("AsOfDate")], "%m/%d/%Y")

# sort: # http://www.statmethods.net/management/sorting.html
portfolios.sorted <- portfolios.dirty[order(portfolios.dirty$AsOfDate, decreasing = FALSE),]
head(portfolios.sorted)

# remove nulls
# NA in this module



#################################
# Estimating Historical Risk
#################################

# new data, preserving original
returns <- portfolios.sorted

# calculate returns
returns[-nrow(returns), -1] <- returns[-nrow(returns), -1] / returns[-1, -1] - 1
head(returns)

# removes last row, since it is not a valid return value
tail(returns)
returns <- returns[-nrow(returns),]
tail(returns)

# create a data frame of stock names, not the date and not the factors
stockNames = names(returns)[-which(names(returns) %in% c('AsOfDate', 'FVX', 'SP500'))]

# create a data frame of only stocks
stockReturns <- returns[stockNames]

# create a data frame of weights, but make it equal weighted
# it is hterfore 1 for each item, divided by the number of names
# in effect, 1 divided by 10
(weights <- data.frame(rep(1, times = length(stockNames)) / length(stockNames)))

# rename the row names, then rename the column header
row.names(weights) <- stockNames
names(weights) <- "Weights"

# cacluate portfolio variance
# Var(p) = W * COV(Y) * W_transposed
# matrix multiplication is done via %*%
# first is transpose of weights
# second is the covariance of the returns
# third is the matrix of weights
(historicalReturnRisk <- t(weights) %*% cov(stockReturns) %*% as.matrix(weights))



#################################
# Building Factor Models
#################################

# calculate lm (regression) for FVX, SP500, and the stock value


# loops for each stock to create
# create data frame
modelCoeffs <- data.frame()
for (stockName in stockNames) {
    stockReturn <- as.matrix(returns[stockName])
    sp500 <- as.matrix(returns["SP500"])
    fvx <- as.matrix(returns["FVX"])
    model <- lm(stockReturn ~ sp500 + fvx)
    modelCoeffs <- rbind(modelCoeffs, cbind(coef(model)[1], coef(model)[2], coef(model)[3], sd(resid(model))))
}

# cleanup data frame
row.names(modelCoeffs) <- stockNames
names(modelCoeffs) <- c("Alpha","B_SP500","B_FVX","ResidualVol")
modelCoeffs



#################################
# Idiosyncratic and Systemi Risk
#################################

# systemic risk
# systemicVariance(P) = weights * betas * covariance(factor returns) * betas_transposed * weight_transposed

# create matrix of factors
factorNames <- c("SP500","FVX")
factorReturns <- returns[factorNames]

# guess
#systemicVariance <- t(weights) %*% as.matrix(weights) %*% cov(factorReturns) %*% t(modelCoeffs) %*% as.matrix(modelCoeffs)

# actual
# internal covariance matrix
(reconstructedCov <- as.matrix(modelCoeffs[c("B_SP500", "B_FVX")]) %*% as.matrix(cov(factorReturns)) %*% t(modelCoeffs[c("B_SP500", "B_FVX")]))
# external covariance matrix
(systemicVariance <- t(weights) %*% reconstructedCov %*% as.matrix(weights))


# idosyncratic risk
# idosyncraticVariance(P) = weights * var(residuals) * weight_transposed
(idosyncraticVariance <- sum((weights ^ 2) * (modelCoeffs$ResidualVol ^ 2)))

(totalvariance = systemicVariance + idosyncraticVariance)



#################################
# Scenario-based Stress Testing
#################################

# currentPrices <- subset(portfolios.sorted, AsOfDate == max(portfolios.sorted$AsOfDate))
(currentPrices <- portfolios.sorted[1,])

# going to use min and max of this to create scenarios
(summary(returns[factorNames]))

(fvxScenarios <- seq(min(returns["FVX"]), max(returns["FVX"]), by = .05))
(sp500Scenarios <- seq(min(returns["SP500"]), max(returns["SP500"]), by = .02))

# build scenarios from fvxScenarios by sp500Scenarios
scenarios <- data.frame()
for (fvxValue in fvxScenarios) {
    for (sp500value in sp500Scenarios) {
        scenario <- cbind(fvxValue, sp500value)
        scenarioPredictedReturns <- data.frame()
        for (stockName in stockNames) {
            alpha = subset(modelCoeffs, row.names(modelCoeffs) %in% c(stockName)) ["Alpha"]
            beta_sp = subset(modelCoeffs, row.names(modelCoeffs) %in% c(stockName))["B_SP500"]
            beta_fvx = subset(modelCoeffs, row.names(modelCoeffs) %in% c(stockName))["B_FVX"]
            scenarioPredictedReturn = alpha + (beta_sp * sp500value) + (beta_fvx * fvxValue)
            scenarioPredictedReturns = rbind(scenarioPredictedReturns, scenarioPredictedReturn)
        }
        scenario <- (cbind(scenario, t(scenarioPredictedReturns)))
        scenarios <- rbind(scenarios, scenario)
    }
}
head(scenarios)


# totalVariance(P) = weights * cov(scenarios) * weight_transposed
(scenarioTotalVariance <- t(weights) %*% cov(as.matrix(scenarios[stockNames]))  %*% as.matrix(weights))



#################################
# Quantifying the Worst Case
#################################

# VaR = P x Z_alpha x stdDev
calculateVaR <- function(risk, confidenceLevel, principal = 1, numMonths = 1) {
    vol <- sqrt(risk)
    data.frame(VaR = abs(principal * qnorm(1-confidenceLevel, 0, 1) * vol * sqrt(numMonths))[1])
}

calculateVaR(scenarioTotalVariance, 0.99)

calculateVaR(historicalReturnRisk, 0.99)

calculateVaR(totalvariance, 0.99)


#################################
# Plotting Var
#################################

library(PerformanceAnalytics)

returnsForChart <- returns

row.names(returnsForChart) <- c(returnsForChart$AsOfDate)

chart.VaRSensitivity(returnsForChart[, 2:11, drop=TRUE],
		methods=c("HistoricalVaR", "ModifiedVaR", "GaussianVaR"),
        colorset = bluefocus,
        lwd=2)

chart.RelativePerformance(returnsForChart[, 2:11],
    returnsForChart[, 13],
    main = "Relative Performace vs. Benchmark",
    colorset = c(1:8),
    xaxis = TRUE,
    ylog = FALSE,
    legend.loc = "topright",
    cex.legend = 0.8)










